File: 94_-_Binary_Tree_Inorder_Traversal.cs
/**
 * 94. Binary Tree Inorder Traversal - Easy
 * Given the root of a binary tree, return the inorder traversal of its nodes' values.
 */

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public IList<int> InorderTraversal(TreeNode root) {
        return InorderTraversalRecursive(root);
    }
    
    public IList<int> InorderTraversalIterative(TreeNode root) {
        IList<int> ans = new List<int>();
        Stack<TreeNode> stack = new Stack<TreeNode>();
        TreeNode current = root;
        
        while (current != null || stack.Count > 0) {
            while (current != null) {
                stack.Push(current);
                current = current.left;
            }
            
            current = stack.Pop();
            ans.Add(current.val);
            
            current = current.right;
        }
        
        return ans;
    }
    
    public IList<int> InorderTraversalRecursive(TreeNode root) {
        IList<int> ans = new List<int>();
        RecursiveHelper(root, ans);
        return ans;
    }
    
    public void RecursiveHelper(TreeNode current, IList<int> list) {
        if (current == null) {
            return;
        }
        
        RecursiveHelper(current.left, list);
        list.Add(current.val);
        RecursiveHelper(current.right, list);
    }
}

File: javascript/94_-_Binary_Tree_Inorder_Traversal.js
/**
 * 94. Binary Tree Inorder Traversal - Easy
 * Given the root of a binary tree, return the inorder traversal of its nodes' values.
 */

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var inorderTraversal = function(root) {
    return inorderTraversalIterative(root);
};

var inorderTraversalIterative = function(root) {
    const ans = [];
    const stack = [];
    let current = root;

    while (current || stack.length > 0) {
        while (current) {
            stack.push(current);
            current = current.left;
        }

        current = stack.pop();
        ans.push(current.val);

        current = current.right;
    }
};

var inorderTraversalRecursive = function(root) {
    const ans = [];
    recursiveHelper(root, ans);
    return ans;
};

var recursiveHelper = function(current, list) {
    if (!current) {
        return;
    }

    recursiveHelper(current.left, list);
    list.push(current.val);
    recursiveHelper(current.right, list);
};